source/frontend-testing.rst

```diff
@@ -0,0 +1,334 @@
+It will check your tests' coverage of statements, branches, functions, and lines found in
+the ``app/`` subdir of the project's JS dir (i.e. in ``project_name/static/js/app/``).
+
+You can set the desired coverage thresholds inside the ``gulpfile.js``. The
+numbers here represent percentages::
+
+    .pipe(coverageEnforcer({
+      thresholds: {
+        statements: 80
+        , branches: 50
+        , lines: 80
+        , functions: 50
+      }
+      // ...
+    }))
+
+
+Testing our stack
+#################
+
+The following front-end tools are bundled in the project template and represents
+our basic stack for future projects:
+
+* `gulp <http://gulpjs.com/>`_, the Node-based streaming build tool
+* `ECMAScript 2015 <http://es6-features.org/>`_, transpiled into JavaScript
+  with `Babel <https://babeljs.io/>`_
+* `React <https://facebook.github.io/react/>`_, a library covering roughly the
+  "view" and "controller" aspects of a MVC-architected user interface
+    * `JSX <https://facebook.github.io/react/docs/jsx-in-depth.html>`_, a JS syntax
+      extension that makes it easier to create React components
+* `Flux <https://facebook.github.io/flux/>`_, a library and design pattern for
+  managing UI application state and building UIs in a "data flow" style
+    * `Immutable.js <https://facebook.github.io/immutable-js/>`_, an immutable
+      data structure library used under the hood in Flux's utilities
+
+This section of the documentation will walk through the process of setting up
+tests that take these technologies into account:
+
+* the basics of using Gulp and Babel to write and run Mocha tests in ES2015
+* special techniques and gotchas related to writing tests for React components and Flux
+  applications
+
+ES2015 Mocha tests with Gulp and Babel
+--------------------------------------
+
+Transpiling ES2015 code is already a standard part of Caktus's front-end process.
+We can take advantage of the hard work that goes into this code preprocessing
+to reuse parts of the process and layer new steps in without serious difficulty.
+
+This comes out clearly in the actual test command, handled by Gulp, which
+transpiles our code using Babel behind the scenes. Our Gulp test task begins::
+
+    gulp.task('test', function () {
+        require('babel-core/register');
+        // ...
+    });
+
+``babel-core/register``, when imported, causes all imports within the scope to
+be run through Babel. The result is that ES2015 and JSX files used by the Mocha
+test runner are preprocessed without our needing to do anything special.
+
+This is guaranteed because our ``.babelrc`` file in the top level dir of the
+project sets up Babel to transpile ES2015 and JSX::
+
+    {
+      "presets": ["es2015"],
+      "plugins": ["transform-react-jsx"]
+    }
+
+This allows us to not only write and test ES2015 and JSX applications but to write
+our tests themselves in ES2015 and JSX.
+
+React tests with jsdom
+----------------------
```

Comment:
Let's include in this section that the principles can apply to other frontend libraries, not just React.